consensus/ibft/src/test/java/tech/pegasys/pantheon/consensus/ibft/validation/NewRoundMessageValidatorTest.java

```diff
@@ -0,0 +1,408 @@
+import tech.pegasys.pantheon.crypto.SECP256K1.KeyPair;
+import tech.pegasys.pantheon.ethereum.core.Address;
+import tech.pegasys.pantheon.ethereum.core.Block;
+import tech.pegasys.pantheon.ethereum.core.Hash;
+import tech.pegasys.pantheon.ethereum.core.Util;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+
+import com.google.common.collect.Lists;
+import org.junit.Before;
+import org.junit.Test;
+
+public class NewRoundMessageValidatorTest {
+
+  private final KeyPair proposerKey = KeyPair.generate();
+  private final KeyPair validatorKey = KeyPair.generate();
+  private final KeyPair otherValidatorKey = KeyPair.generate();
+  private final MessageFactory proposerMessageFactory = new MessageFactory(proposerKey);
+  private final MessageFactory validatorMessageFactory = new MessageFactory(validatorKey);
+  private final MessageFactory otherValidatorMessageFactory = new MessageFactory(otherValidatorKey);
+  private final Address proposerAdress = Util.publicKeyToAddress(proposerKey.getPublicKey());
+
+  private final List<Address> validators = Lists.newArrayList();
+
+  private final Block proposedBlock = mock(Block.class);
+  private final ProposerSelector proposerSelector = mock(ProposerSelector.class);
+  private final MessageValidatorFactory validatorFactory = mock(MessageValidatorFactory.class);
+  private final MessageValidator messageValidator = mock(MessageValidator.class);
+  private final ConsensusRoundIdentifier roundIdentifier = new ConsensusRoundIdentifier(2, 4);
+  private final RoundChangeCertificate validCert = mock(RoundChangeCertificate.class);
+
+  @Before
+  public void setup() {
+    validators.add(Util.publicKeyToAddress(proposerKey.getPublicKey()));
+    validators.add(Util.publicKeyToAddress(validatorKey.getPublicKey()));
+    validators.add(Util.publicKeyToAddress(otherValidatorKey.getPublicKey()));
+
+    when(proposerSelector.selectProposerForRound(any())).thenReturn(proposerAdress);
+
+    when(validatorFactory.createAt(any())).thenReturn(messageValidator);
+    when(messageValidator.addSignedProposalPayload(any())).thenReturn(true);
+    when(messageValidator.validatePrepareMessage(any())).thenReturn(true);
+
+    when(proposedBlock.getHash()).thenReturn(Hash.fromHexStringLenient("1"));
+  }
+
+  @Test
+  public void newRoundFromNonProposerFails() {
+    final NewRoundMessageValidator validator =
+        new NewRoundMessageValidator(
+            validators, proposerSelector, validatorFactory, 1, roundIdentifier);
+
+    final SignedData<NewRoundPayload> msg =
+        validatorMessageFactory.createSignedNewRoundPayload(
+            roundIdentifier,
+            new RoundChangeCertificate(Collections.emptyList()),
+            proposerMessageFactory.createSignedProposalPayload(roundIdentifier, proposedBlock));
+
+    assertThat(validator.validateNewRoundMessage(msg)).isFalse();
```

Comment:
This test would succeed even if the message is signed by the proposer.

Reply:
done.